### PR TITLE
[5.0][Events] Application configuration event classes

### PR DESCRIPTION
**Pull request by Fedik**
### Summary of Changes
Implementing Event classes for Application configuration events.
This also need https://github.com/joomla/joomla-cms/pull/41226, that has some fixes. But can be tested without.

Affected events:

- [x] onApplicationBeforeSave
- [x] onApplicationAfterSave

### Testing Instructions
Apply patch.
Try edit global  configuration.
All should work as before.

### Actual result BEFORE applying this Pull Request
Works

### Expected result AFTER applying this Pull Request
Works

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: https://docs.joomla.org/Plugin/Events#Config_Component
- [ ] No documentation changes for docs.joomla.org needed
- [ ] Pull Request link for manual.joomla.org: TBD
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: libraries/src/Event/Application/ApplicationConfigurationEvent.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function __construct($name, array $arguments = [])
+    {
+        parent::__construct($name, $arguments);
+
+        if (!\array_key_exists('subject', $this->arguments)) {
+            throw new \BadMethodCallException("Argument 'subject' of event {$name} is required but has not been provided");
+        }
```
Comment (heelc29) [inline libraries/src/Event/Application/ApplicationConfigurationEvent.php:42]: In the other events the check of arguments is before the parent construct...
```suggestion
        if (!\array_key_exists('subject', $this->arguments)) {
            throw new \BadMethodCallException("Argument 'subject' of event {$name} is required but has not been provided");
        }

        parent::__construct($name, $arguments);
```

File: administrator/components/com_config/src/Model/ApplicationModel.php [hunk 2]
```diff
@@ -744,7 +746,9 @@ public function save($data)
         // Clear cache of com_config component.
         $this->cleanCache('_system');
 
-        $result = $app->triggerEvent('onApplicationBeforeSave', [$config]);
+        $eventBefore = new BeforeSaveConfigurationEvent('onApplicationBeforeSave', ['subject' => $config]);
+        $this->getDispatcher()->dispatch('onApplicationBeforeSave', $eventBefore);
+        $result = $eventBefore['result'] ?? [];
```
Comment (heelc29) [inline administrator/components/com_config/src/Model/ApplicationModel.php]: also an option?
```suggestion
        $result = $eventBefore->getArgument('result', []);
```

**Comment by heelc29** [comment]
There is a small b/c break
image
```diff
- $result = $app->triggerEvent('onApplicationBeforeSave', [$config]);
+ $result = $app->triggerEvent('onApplicationBeforeSave', ['subject' => $config]);
```

image
```diff
- $app->triggerEvent('onApplicationAfterSave', [$config]);
+ $app->triggerEvent('onApplicationAfterSave', ['subject' => $config]);
```

you will see it when you perform `removeroot`
https://github.com/joomla/joomla-cms/blob/fe2074e891b404ccb2922b5932cb9beaec894ebf/administrator/components/com_config/src/Model/ApplicationModel.php#L773-L799

**Comment by heelc29** [comment]
I have tested this item :red_circle: unsuccessfully on 63854ecf90ecc80b53699f715bb77554cfb026dd<br /><br />See comments above<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41325">issues.joomla.org/tracker/joomla-cms/41325</a>.</sub>

File: libraries/src/Event/Application/ApplicationConfigurationEvent.php (hunk 1 above)
Comment (Fedik) [reply libraries/src/Event/Application/ApplicationConfigurationEvent.php:42]: I changed to be after, also in other my PR, this have own benefits. Example when child class may have defined arguments already.

File: administrator/components/com_config/src/Model/ApplicationModel.php (hunk 2 above)
Comment (Fedik) [reply administrator/components/com_config/src/Model/ApplicationModel.php]: Both is okay, but I already use first one everywhere :)

Comment (Fedik) [reply administrator/components/com_config/src/Model/ApplicationModel.php]: Maybe later I will review it again.

**Comment by Fedik** [comment]
Yea, I have missed couple ones, should be good now

File: administrator/components/com_config/src/Model/ApplicationModel.php (hunk 2 above)
Comment (Fedik) [reply administrator/components/com_config/src/Model/ApplicationModel.php]: I added getArgument, in a bit diferent way :)

**Comment by heelc29** [comment]
I have tested this item :white_check_mark: successfully on bb8986c4c325ec2bafda1a0c02ea25870a1d73b2<br /><br />Looks good now. Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41325">issues.joomla.org/tracker/joomla-cms/41325</a>.</sub>

**Comment by HLeithner** [comment]
Merging this for the next alpha to get more feedback. Thanks for the work.

**Comment by HLeithner** [comment]
thanks

**Pull request merged**